dotCMS/src/main/java/com/dotmarketing/services/PageServices.java

```diff
@@ -197,126 +197,142 @@ public static InputStream buildStream(IHTMLPage htmlPage, Identifier identifier,
+						if (EDIT_MODE)
+							contentletsFull = contentlets;
+						else
+							contentletsFull =
+								conAPI.findPageContentlets(idenHtmlPage.getId(), idenContainer.getId(), sort, true, -1,
+									APILocator.getUserAPI().getSystemUser(), false);
+					} catch (Exception e) {
+						Logger.error(PageServices.class, "Unable to retrive contentlets on page", e);
+					}
+					Logger.debug(PageServices.class,
+						"HTMLPage= " + htmlPage.getInode() + " Container=" + c.getInode() + " Language=-1 Contentlets="
+							+ contentlets.size());
+				}
+				//this is to filter the contentlets list removing the repited identifiers
+				if (contentlets.size() > 0) {
+					Set<String> contentletIdentList = new HashSet<String>();
+					List<Contentlet> contentletsFilter = new ArrayList<Contentlet>();
+					for (Contentlet cont : contentlets) {
+						if (!contentletIdentList.contains(cont.getIdentifier())) {
+							contentletIdentList.add(cont.getIdentifier());
+							contentletsFilter.add(cont);
+						}
+					}
+					contentlets = contentletsFilter;
+				}
+				if (contentletsFull.size() > 0) {
+					Set<String> contentletIdentList = new HashSet<String>();
+					List<Contentlet> contentletsFilter = new ArrayList<Contentlet>();
+					for (Contentlet cont : contentletsFull) {
+						if (!contentletIdentList.contains(cont.getIdentifier())) {
+							contentletIdentList.add(cont.getIdentifier());
+							contentletsFilter.add(cont);
+						}
+					}
+					contentletsFull = contentletsFilter;
+				}
 
-			List<Contentlet> contentlets = new ArrayList<Contentlet>();
-			List<Contentlet> contentletsFull = new ArrayList<Contentlet>();
-			if (!dynamicContainer) {
-				Identifier idenHtmlPage = APILocator.getIdentifierAPI().find(htmlPage);
-				Identifier idenContainer = APILocator.getIdentifierAPI().find(c);
-				//The container doesn't have categories
-				try{
-					contentlets = conAPI.findPageContentlets(idenHtmlPage.getId(), idenContainer.getId(), sort, EDIT_MODE, -1,APILocator.getUserAPI().getSystemUser() ,false);
-					if(EDIT_MODE)
-					    contentletsFull=contentlets;
-					else
-					    contentletsFull = conAPI.findPageContentlets(idenHtmlPage.getId(), idenContainer.getId(), sort, true, -1,APILocator.getUserAPI().getSystemUser() ,false);
-				}catch(Exception e){
-					Logger.error(PageServices.class,"Unable to retrive contentlets on page", e);
+				StringBuilder widgetpree = new StringBuilder();
+				StringBuilder widgetpreeFull = new StringBuilder();
+
+				StringBuilder contentletList = new StringBuilder();
+				int count = 0;
+				for (Contentlet contentlet : contentlets) {
+					contentletList.append(count == 0 ? "" : ",")
+						.append('"').append(contentlet.getIdentifier()).append('"');
+					if (contentlet.getStructure().getStructureType() == Structure.STRUCTURE_TYPE_WIDGET) {
+						Field field = contentlet.getStructure().getFieldVar("widgetPreexecute");
+						if (field != null && UtilMethods.isSet(field.getValues()))
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 4 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS134)